File: 08042021/bfs-graph.js
// BreadthFirstSeach

const bfsPrint = (graph, source) => {
    const queue = [source];
    while (queue.length > 0) {
        const curr = queue.shift();
        console.log(curr);
        for (let neib of graph[curr]) {
            queue.push(neib);
        }
    }
};

const graph = {
    a: ["b", "c"],
    b: ["d"],
    c: ["e"],
    d: ["f"],
    e: [],
    f: [],
};

bfsPrint(graph, "a");

File: 08042021/bfs-graph.cpp
// Breadth First Search

#include <iostream>
#include <vector>
#include <map>
#include <queue>
using namespace std;

void bfsPrint(map<char, vector<char>>& graph, char src)
{
	queue<char> que;
	que.push(src);
	while(!que.empty())
	{
		const char curr = que.front();
		que.pop();
		cout << curr << "->";
		for(char neib : graph[curr])
		{
			que.push(neib);
		}
	}
}


int main()
{
	map<char, vector<char>> graph;
	graph['a'] = {'b', 'c'};
	graph['b'] = {'d'};
	graph['c'] = {'e'};
	graph['d'] = {'f'};
	graph['e'] = {};
	graph['f'] = {};
	
	bfsPrint(graph, 'a');
	cout << endl;
}